File: ios/ExpoThemeChangerModule.swift
import ExpoModulesCore
import UIKit

public class ExpoThemeChangerModule: Module {
  private var lastSystemTheme: String?
  
  public func definition() -> ModuleDefinition {
    Name("ExpoThemeChanger")

    Events("onChangeTheme")

    OnCreate {
      self.lastSystemTheme = self.getCurrentSystemTheme()
      self.startObservingTraitChanges()
    }

    OnDestroy {
      self.stopObservingTraitChanges()
      self.lastSystemTheme = nil
    }

    OnAppEntersForeground {
      self.checkSystemThemeChange()
    }

    Function("setTheme") { (theme: String) -> Void in
      // Validate theme value
      let validThemes = ["light", "dark", "system"]
      guard validThemes.contains(theme) else {
        throw NSError(domain: "ExpoThemeChanger", code: 1, 
                     userInfo: [NSLocalizedDescriptionKey: "Theme must be 'light', 'dark', or 'system'"])
      }
      
      UserDefaults.standard.set(theme, forKey: "theme")
      self.sendEvent("onChangeTheme", [
        "theme": theme,
        "effectiveTheme": self.getEffectiveTheme(theme: theme)
      ])
    }
    
    Function("getTheme") { () -> String in
      UserDefaults.standard.string(forKey: "theme") ?? "system"
    }

    Function("getEffectiveTheme") { () -> String in
      let savedTheme = UserDefaults.standard.string(forKey: "theme") ?? "system"
      return self.getEffectiveTheme(theme: savedTheme)
    }

    Function("getSystemTheme") { () -> String in
      return self.getCurrentSystemTheme()
    }
  }

  private func getCurrentSystemTheme() -> String {
    if #available(iOS 13.0, *) {
      // Get the key window's trait collection for accurate theme detection
      if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
         let window = windowScene.windows.first {
        switch window.traitCollection.userInterfaceStyle {
        case .dark:
          return "dark"
        case .light:
          return "light"
        default:
          return "light"
        }
      }
      // Fallback to UITraitCollection.current
      switch UITraitCollection.current.userInterfaceStyle {
      case .dark:
        return "dark"
      case .light:
        return "light"
      default:
        return "light"
      }
    } else {
      return "light"
    }
  }

  private func getEffectiveTheme(theme: String) -> String {
    if theme == "system" {
      return getCurrentSystemTheme()
    } else {
      return theme
    }
  }

  private func startObservingTraitChanges() {
    if #available(iOS 13.0, *) {
      NotificationCenter.default.addObserver(
        forName: UIScene.didActivateNotification,
        object: nil,
        queue: .main
      ) { [weak self] _ in
        self?.checkSystemThemeChange()
      }
    }
  }

  private func stopObservingTraitChanges() {
    NotificationCenter.default.removeObserver(self)
  }

  private func checkSystemThemeChange() {
    let currentSystemTheme = getCurrentSystemTheme()
    let savedTheme = UserDefaults.standard.string(forKey: "theme") ?? "system"
    
    // Only emit event if theme is set to "system" and system theme actually changed
    if savedTheme == "system" && lastSystemTheme != currentSystemTheme {
      lastSystemTheme = currentSystemTheme
      sendEvent("onChangeTheme", [
        "theme": "system",
        "effectiveTheme": currentSystemTheme
      ])
    } else {
      lastSystemTheme = currentSystemTheme
    }
  }
}

File: android/src/main/java/expo/modules/themechanger/ExpoThemeChangerModule.kt
package expo.modules.themechanger

import android.content.Context
import android.content.res.Configuration
import androidx.core.os.bundleOf
import android.content.SharedPreferences
import android.content.ComponentCallbacks2
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoThemeChangerModule : Module() {
  private var lastSystemTheme: String? = null
  private var configCallback: ComponentCallbacks2? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoThemeChanger")

    Events("onChangeTheme")

    OnCreate {
      lastSystemTheme = getCurrentSystemTheme()
      registerConfigurationCallback()
    }

    OnDestroy {
      unregisterConfigurationCallback()
      lastSystemTheme = null
    }

    OnActivityEntersForeground {
      // Check if system theme changed while app was in background
      checkSystemThemeChange()
    }

    Function("setTheme") { theme: String ->
      // Validate theme value
      if (theme !in listOf("light", "dark", "system")) {
        throw IllegalArgumentException("Theme must be 'light', 'dark', or 'system'")
      }
      
      getPreferences().edit().putString("theme", theme).apply()
      this@ExpoThemeChangerModule.sendEvent("onChangeTheme", bundleOf(
        "theme" to theme,
        "effectiveTheme" to getEffectiveTheme(theme)
      ))
    }

    Function("getTheme") {
      return@Function getPreferences().getString("theme", "system")
    }

    Function("getEffectiveTheme") {
      val savedTheme = getPreferences().getString("theme", "system") ?: "system"
      return@Function getEffectiveTheme(savedTheme)
    }

    Function("getSystemTheme") {
      return@Function getCurrentSystemTheme()
    }
  }

  private val context
    get() = requireNotNull(appContext.reactContext)

  private fun getPreferences(): SharedPreferences {
    return context.getSharedPreferences(context.packageName + ".settings", Context.MODE_PRIVATE)
  }

  private fun getCurrentSystemTheme(): String {
    val nightModeFlags = context.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
    return when (nightModeFlags) {
      Configuration.UI_MODE_NIGHT_YES -> "dark"
      Configuration.UI_MODE_NIGHT_NO -> "light"
      else -> "light"
    }
  }

  private fun getEffectiveTheme(theme: String): String {
    return if (theme == "system") {
      getCurrentSystemTheme()
    } else {
      theme
    }
  }

  private fun registerConfigurationCallback() {
    configCallback = object : ComponentCallbacks2 {
      override fun onConfigurationChanged(newConfig: Configuration) {
        checkSystemThemeChange()
      }

      override fun onLowMemory() {}
      override fun onTrimMemory(level: Int) {}
    }
    
    try {
      context.registerComponentCallbacks(configCallback)
    } catch (e: Exception) {
      // Context might not be ready yet
    }
  }

  private fun unregisterConfigurationCallback() {
    configCallback?.let {
      try {
        context.unregisterComponentCallbacks(it)
      } catch (e: Exception) {
        // Context might already be destroyed
      }
    }
    configCallback = null
  }

  private fun checkSystemThemeChange() {
    val currentSystemTheme = getCurrentSystemTheme()
    val savedTheme = getPreferences().getString("theme", "system") ?: "system"
    
    // Only emit event if theme is set to "system" and system theme actually changed
    if (savedTheme == "system" && lastSystemTheme != currentSystemTheme) {
      lastSystemTheme = currentSystemTheme
      this@ExpoThemeChangerModule.sendEvent("onChangeTheme", bundleOf(
        "theme" to "system",
        "effectiveTheme" to currentSystemTheme
      ))
    } else {
      lastSystemTheme = currentSystemTheme
    }
  }
}